src/main/java/app/model/agents/WallFollow/WallFollowAgent.java

```diff
@@ -0,0 +1,417 @@
+        for (Ray r : view)
+        {
+            // TODO need to deal with edge case for 0 = 360 degrees and area around that
+            if ((r.angle() <= rayAngle + 1.0 && r.angle() >= rayAngle - 1.0) && r.rayLength() <= moveLength) {
+                if (DEBUG)
+                {
+                    System.out.print("WALL DETECTED! Angle of detecting ray: " + rayAngle + " and ray length: ");
+                    System.out.println(r.rayLength());
+                }
+                return false;
+            }
+        }
+        if (DEBUG)
+        {
+            System.out.println("No wall detected with ray of angle: " + rayAngle);
+        }
+        return true;
+    }
+
+    public boolean noObstacleDetectedFromPos(double rayAngle, int distFromAgentVertex)
+    {
+        // use this obstacle detection method for checking for obstacles from vertices that are not the agent's current location
+        //double posDiff = position.sub(vertexPos).length();
+        // distFromAgentVertex counts how many vertices apart are vertex to be checked for obstacle and agent position vertex
+        // e.g. neighbour vertices have distance 1, next level is 2 and so on
+        // if the ray with given rayAngle (+- 1.0 degrees) detects an obstacle in the moveLength distance range -> return false
+        for (Ray r : view)
+        {
+            if (r.angle() <= rayAngle + 1.0 && r.angle() >= rayAngle - 1.0)
+            {
+                if (r.rayLength() <= moveLength*distFromAgentVertex)
+                {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    public boolean obstacleDetected(Vector direction)
+    {
+        // TODO use BoundaryImp methods to check if ray finds obstacle
+        return false;
+    }
+
+    public Vector rotateAgentLeft(boolean rotateLeft)
+    {
+        // if rotateLeft is true, rotate agent 90 degrees left
+        // if rotateLeft id false, rotate agent 90 degrees right
+        List<Vector> directions = Arrays.asList(new Vector(0,1),
+                new Vector(1,0),
+                new Vector(0,-1),
+                new Vector(-1,0));
+        for (int i=0; i<=directions.size(); i++)
+        {
+            if (direction.equals(directions.get(i)) && !rotateLeft)
+            {
+                if (i == 0)
+                {
+                    return directions.get(3);
+                }
+                else
+                {
+                    return directions.get(i-1);
+                }
+            }
+            else if (direction.equals(directions.get(i)) && rotateLeft)
+            {
+                if (i == directions.size()-1)
+                {
+                    return directions.get(0);
+                }
+                else
+                {
+                    return directions.get(i+1);
+                }
+            }
+        }
+        // should never reach here when only 4 directions are possible
```

Comment:
Could throw an error here if this does get reached, that way you find out if you have an issue
```java
throw new RuntimeException("All four directions ...");
```